src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+                    }
+                    (None, Some(_)) => unreachable!(),
+                },
+                Message::Data(Update {
+                    row,
+                    timestamp,
+                    diff,
+                }) => {
+                    let time = if let Some(frontier) = compaction_frontier {
+                        std::cmp::min(frontier, *timestamp)
+                    } else {
+                        *timestamp
+                    };
+                    let entry = time_data.entry((time, row)).or_insert(0);
+                    *entry += diff;
+                }
+            }
+        }
+
+        // Now let's prepare the output
+        let mut buf = Vec::new();
+        assert!(lower.is_some());
+        assert!(upper.is_some());
+
+        let lower = lower.unwrap();
+        let upper = upper.unwrap();
+
+        // Frame each batch with its lower and upper bound timestamp.
+        // TODO: match the behavior of CDCv2 updates with a count of messages
+        // at each timestamp.
+        encode_progress(lower, &mut buf)?;
+        for ((timestamp, row), diff) in time_data.into_iter() {
+            if diff == 0 {
+                continue;
+            }
+
+            encode_update(row, timestamp, diff, &mut buf)?;
+        }
+
+        encode_progress(upper, &mut buf)?;
+
+        let batch_name = format!("batch-{}-{}", lower, upper);
+        let batch_path = trace_path.join(&batch_name);
+        let batch_tmp_path = trace_path.join(format!("{}-tmp", batch_name));
+        // Write the file first suffixed with "-tmp" and then rename to guard against
+        // partial writes.
+        fs::write(&batch_tmp_path, buf)
+            .with_context(|| format!("failed to write batch file {}", batch_tmp_path.display()))?;
+        fs::rename(&batch_tmp_path, &batch_path).with_context(|| {
+            format!(
+                "failed to rename batch file from: {} to: {}",
+                batch_tmp_path.display(),
+                batch_path.display()
+            )
+        })?;
+
+        Ok(Batch {
+            upper,
+            lower,
+            path: batch_path,
+        })
+    }
+
+    /// Reintroduce a batch based on an available file in the trace
+    /// directory.
+    fn reinit(path: PathBuf) -> Result<Self, anyhow::Error> {
+        let batch_name = path
+            .file_name()
+            .expect("batch name known to exist")
+            .to_str()
+            .expect("batch name known to be valid utf8");
+        let parts: Vec<_> = batch_name.split('-').collect();
+        assert!(parts.len() == 3);
+        Ok(Self {
+            upper: parts[2].parse()?,
+            lower: parts[1].parse()?,
+            path,
+        })
+    }
+
+    /// Read the data from a batch stored on disk into memory.
+    fn read(&self) -> Result<Vec<Message>, anyhow::Error> {
+        read_segment(&self.path)
+    }
+
+    /// Physically concatenate all batches together into a single batch
```

Comment:
maybe replace "all batches" with `batches`? The text makes me wonder what if `batches` isn't all batches, or if that is really what you mean (I haven't read the method yet, so just trying to square the doc comments with the signature).